pennylane_lightning/src/gates/MyGateImplementation.hpp

```diff
@@ -0,0 +1,160 @@
+#include <vector>
+#include <thread>
+#include <algorithm>
+#include "BitUtil.hpp"
+#include "GateUtil.hpp"
+#include "Gates.hpp"
+#include "GateOperation.hpp"
+#include "KernelType.hpp"
+#include "PauliGenerator.hpp"
+
+using namespace Pennylane;
+using namespace Pennylane::Gates;
+
+class MyGateImplementation: public PauliGenerator<MyGateImplementation>{
+  public:
+    constexpr static KernelType kernel_id = KernelType::MyKernel; // Will be discussed below
+    constexpr static std::string_view name = "MyGateImpl"; // Name of your kernel
+
+    constexpr static std::array implemented_gates = {
+        GateOperation::PauliX,
+        GateOperation::PauliZ
+    }; // List of implemented gates
+
+
+    [[maybe_unused]] constexpr static std::array<GeneratorOperation,1> implemented_generators{};
+    
+    static inline 
+    std::vector<size_t> bounds(size_t parts, size_t mem){
+         std::vector<size_t> bnd;
+        size_t delta = mem/parts;
+        size_t reminder = mem%parts;
+        size_t N1=0, N2=0;
+        bnd.push_back(N1);
+        for(size_t i=0;i<parts;++i){
+            N2 = N1 + delta;
+            if(i == parts -1)
+                 N2+=reminder;
+            bnd.push_back(N2);
+            N1 = N2;
+        }
+        return bnd;
+    }
+
+
+    template <class PrecisionT>
+    static inline void PauliX_thread(std::complex<PrecisionT>* data, size_t L, size_t R, size_t wire_parity_inv, size_t wire_parity, size_t rev_wire_shift){
+        
+        size_t n0;
+        size_t i00;
+        size_t i10;
+    
+        for(size_t k=L; k<R;k=k+2){
+            #pragma unroll
```

Comment:
What advantage is there for adding loop unrolling here?

Reply:
Also, this fails to be recognised on our CI tests as a valid option; we compile using clang-tidy to test and with warnings enabled as errors.

Reply:
Launching/destroying threads needs time and the time spent on it is much longer than the time reduced on the qubit operation by using threads when the qubit counts is low. When the qubit counts is larger, the overhead caused by launching/destroying threads will be shorter than the time reduced on the qubit operation by using threads when the qubit counts is low.  Thanks @chaeyeunpark